### ISSUE TITLE
Test program broke `mingw-x86.cmake` on Ubuntu Bionic due to `fmt`. (bisected)

### ISSUE THREAD

**Issue by lss4** [maharmstone/btrfs#522]
**TL;DR:** The test part in `CMakeLists.txt` introduced `fmt` which broke `cmake`. Turning off `WITH_TEST` in `CMakeLists.txt` will allow `cmake` to complete so I could proceed with building.

This applies to Ubuntu Bionic, and after some further testing, Focal as well. Looks like both Bionic and Focal can produce `shellbtrfs.dll` that can be used on XP/2003.

---

A follow up of my old issue #378.

As hinted in [here](https://github.com/maharmstone/btrfs/issues/508#issuecomment-1233310146), it seems `shellbtrfs.dll` still does not work and the following imports not present on XP can be found with Dependency Walker.

```
CompareStringEx
GetLocaleInfoEx
InitializeCriticalSectionEx
LCMapStringEx
```

Both debug and release versions import those. Turned out these are the same few imports I mentioned in my previous issue...

I tried building the latest commit of this particular part on my Ubuntu Bionic VM to get a compatible output like previous versions, but this time I'm failing on `cmake` (I've the most recent version of it installed from Kitware repo, though the snap one will also work).

It first complains about the absence of the `fmt` library, which can be addressed by installing `libfmt-dev:i386` via `apt-get` (this will remove the 64-bit counterpart, `libfmt-dev` as the two cannot coexist).

After satisfying the `fmt` requirement I've now hit a wall with it requiring `CXX20` due to `fmt::fmt-header-only`. Of course, Bionic's GCC is too old to support this.

I've bisected the first commit that introduced the `fmt` which broke `cmake` on Bionic: 67af5c5430a14c825d1d7bb2d7f8b33e74d723b4

Additionally, for `cmake` to complete successfully on commits before this one, this commit (df5de5f8a812d06fa399d043ad3f5c605c204bf9) needs to be cherry-picked, to point to the correct compiler.

**Comment by lss4** [comment]
Sorry, pressed the create issue button by accident so I edited the OP to add the part that I didn't finish.

It seems the `fmt` might have been introduced by test related stuffs... will see if it's possible to omit that part...

**Comment by lss4** [comment]
Okay I've figured it out myself. I indeed need to disable test. After that I could produce a usable `shellbtrfs.dll` that can work on XP/2003 (no missing symbols).

`option(WITH_TEST "Compile test program" ON)`

Change the ON to OFF and `cmake` will be able to complete normally.

PS: My own commit at that time e44152c040ae725b239df38e0dd4e1fe8f6bc627 (which removes an excessive `/usr` in the path) is still needed when building on Bionic. Maybe I can consider bringing up a CentOS VM for building `shellbtrfs` if the path is indeed correct there...

**Comment by maharmstone** [comment]
> In CMakeLists.txt:
>```
> option(WITH_TEST "Compile test program" ON)
>```
> Change the ON to OFF and cmake will be able to complete normally.

That'll work, though the normal way of doing it is `cmake -DWITH_TEST=OFF`.

> PS: My own commit at that time https://github.com/maharmstone/btrfs/commit/e44152c040ae725b239df38e0dd4e1fe8f6bc627 (which removes an excessive /usr in the path) is still needed when building on Ubuntu (Bionic/Focal).

There's probably a better variable I could have used here rather than CMAKE_FIND_ROOT_PATH.

> Maybe I can consider bringing up a CentOS VM for building shellbtrfs if the path is indeed correct there...

Or Docker would be easier still, I think. If it helps, I'm using gcc 12.2 and mingw64-runtime 10. I'm going to close this issue now, as I'm not going to patch the repo so it compiles on old toolchains.
